46-lightning-detector/firmware/lightning_detector/lightning_detector.ino

```diff
@@ -0,0 +1,456 @@
+    if (req == NULL) {
+        Serial.println("Failed to create note.template request.");
+        return;
+    }
+
+    J *body = JCreateObject();
+    if (body == NULL) {
+        JDelete(req);
+        Serial.println("Failed to create note.template request body.");
+        return;
+    }
+
+    JAddStringToObject(req, "file", EVENTS_FILE);
+    JAddNumberToObject(body, TYPE_FIELD, TUINT8);
+    JAddNumberToObject(body, DIST_TO_STORM_FIELD, TINT8);
+    JAddNumberToObject(body, ENERGY_FIELD, TUINT32);
+    JAddItemToObject(req, "body", body);
+
+    if (!notecard.sendRequest(req)) {
+        Serial.println("Failed to send note.template request.");
+    }
+}
+
+void NoteUserAgentUpdate(J *ua)
+{
+    JAddStringToObject(ua, "app", "nf46");
+}
+
+// Convert the binary distance value from the sensor to kilometers. The mappings
+// here come from the sensor's datasheet, Figure 42: Distance Estimation.
+// https://www.mouser.com/datasheet/2/588/ams_AS3935_Datasheet_EN_v5-1214568.pdf
+static int8_t decodeDistanceValue(uint8_t binaryEncodedDist)
+{
+    int8_t distance = DISTANCE_OUT_OF_RANGE;
+
+    switch (binaryEncodedDist) {
+        case 0x7F:
+            distance = DISTANCE_OUT_OF_RANGE;
+            break;
+        case 0x28:
+            distance = 40;
+            break;
+        case 0x25:
+            distance = 37;
+            break;
+        case 0x22:
+            distance = 34;
+            break;
+        case 0x1F:
+            distance = 31;
+            break;
+        case 0x1B:
+            distance = 27;
+            break;
+        case 0x18:
+            distance = 24;
+            break;
+        case 0x14:
+            distance = 20;
+            break;
+        case 0x11:
+            distance = 17;
+            break;
+        case 0x0E:
+            distance = 14;
+            break;
+        case 0x0C:
+            distance = 12;
+            break;
+        case 0x0A:
+            distance = 10;
+            break;
+        case 0x08:
+            distance = 8;
+            break;
+        case 0x06:
+            distance = 6;
+            break;
+        case 0x05:
+            distance = 5;
+            break;
+        case 0x01:
+            distance = 0;
+            break;
+        default:
+            break;
+    }
+
+    return distance;
+}
+
+static void sendEventNote(EventType type, int8_t distance, uint32_t energy)
+{
+    Serial.println("Sending event note...");
+
+    J *req = notecard.newRequest("note.add");
+    if (req == NULL) {
+        Serial.println("Failed to create note.add request.");
+        return;
+    }
+
+    J *body = JCreateObject();
+    if (body == NULL) {
+        JDelete(req);
+        Serial.println("Failed to create note.add request body.");
+        return;
+    }
+
+    JAddStringToObject(req, "file", EVENTS_FILE);
+    JAddNumberToObject(body, TYPE_FIELD, type);
+    JAddNumberToObject(body, DIST_TO_STORM_FIELD, distance);
+    JAddNumberToObject(body, ENERGY_FIELD, energy);
```

Comment:
If the energy and distance fields are omitted when 0 (as described in the README) doesn't that mean that noise events (with type value 0) are also omitted?

Reply:
That's an interesting point. Hadn't thought about it. I can 1-index the types instead of 0-indexing to get around this.